probe/endpoint/reporter.go

```diff
@@ -79,140 +103,126 @@ func (r *Reporter) Report() (report.Report, error) {
 
 	hostNodeID := report.MakeHostNodeID(r.hostID)
 	rpt := report.MakeReport()
+	seenTuples := map[string]fourTuple{}
+
+	// Consult the flowWalker for short-live connections
+	{
+		extraNodeInfo := report.MakeNode().WithLatests(map[string]string{
+			Conntracked: "true",
+		})
+		r.flowWalker.walkFlows(func(f flow) {
+			tuple := fourTuple{
+				f.Original.Layer3.SrcIP,
+				f.Original.Layer3.DstIP,
+				uint16(f.Original.Layer4.SrcPort),
+				uint16(f.Original.Layer4.DstPort),
+			}
+			seenTuples[tuple.key()] = tuple
+			r.addConnection(&rpt, tuple, &extraNodeInfo, &extraNodeInfo)
+		})
+	}
 
 	{
 		conns, err := r.scanner.Connections(r.includeProcesses)
 		if err != nil {
 			return rpt, err
 		}
-		commonNodeInfo := report.MakeNode().WithLatests(map[string]string{
+		toNodeInfo := report.MakeNode().WithLatests(map[string]string{
 			Procspied: "true",
 		})
 		for conn := conns.Next(); conn != nil; conn = conns.Next() {
 			var (
-				localPort  = conn.LocalPort
-				remotePort = conn.RemotePort
-				localAddr  = conn.LocalAddress.String()
-				remoteAddr = conn.RemoteAddress.String()
+				tuple        = fourTuple{conn.LocalAddress.String(), conn.RemoteAddress.String(), conn.LocalPort, conn.RemotePort}
+				fromNodeInfo = toNodeInfo.Copy()
 			)
-			extraNodeInfo := commonNodeInfo.Copy()
 			if conn.Proc.PID > 0 {
-				extraNodeInfo = extraNodeInfo.WithLatests(map[string]string{
+				fromNodeInfo = fromNodeInfo.WithLatests(map[string]string{
 					process.PID:       strconv.FormatUint(uint64(conn.Proc.PID), 10),
 					report.HostNodeID: hostNodeID,
 				})
 			}
-			r.addConnection(&rpt, localAddr, remoteAddr, localPort, remotePort, &extraNodeInfo, &commonNodeInfo)
-		}
-	}
 
-	// Consult the flowWalker for short-live connections
-	{
-		extraNodeInfo := report.MakeNode().WithLatests(map[string]string{
-			Conntracked: "true",
-		})
-		r.flowWalker.walkFlows(func(f flow) {
-			var (
-				localPort  = uint16(f.Original.Layer4.SrcPort)
-				remotePort = uint16(f.Original.Layer4.DstPort)
-				localAddr  = f.Original.Layer3.SrcIP
-				remoteAddr = f.Original.Layer3.DstIP
-			)
-			r.addConnection(&rpt, localAddr, remoteAddr, localPort, remotePort, &extraNodeInfo, &extraNodeInfo)
-		})
+			// If we've already seen this connection, we should know the direction
+			// (or have already figured it out), so we normalize and use the
+			// canonical direction. Otherwise, we can use a port-heuristic to guess
+			// the direction.
+			if canonical, ok := seenTuples[tuple.key()]; ok {
+				if tuple != canonical {
+					tuple = canonical
+					toNodeInfo, fromNodeInfo = fromNodeInfo, toNodeInfo
+				}
+			} else if int(tuple.fromPort) < int(tuple.toPort) {
+				tuple.reverse()
+				toNodeInfo, fromNodeInfo = fromNodeInfo, toNodeInfo
+			}
```

Comment:
I think this can probably be condensed into one if:
```go
canonical, ok := seenTuples[tuple.key()]
if (ok && canonical != tuple) || (!ok && tuple.fromPort < tuple.toPort) {
  tuple.reverse()
  toNodeInfo, fromNodeInfo = fromNodeInfo, toNodeInfo
}
```